app/lib/admin/backend.dart

```diff
@@ -0,0 +1,200 @@
+    return await _withAdmin((user) async {
+      final query = _db.query<User>()
+        ..order('__key__')
+        ..limit(limit);
+
+      if (continuationToken != null) {
+        String lastId;
+        try {
+          lastId = _continuationCodec.decode(continuationToken);
+        } on FormatException catch (_) {
+          throw InvalidInputException.continuationParseError();
+        }
+        InvalidInputException.checkNotNull(lastId, 'continuationToken');
+
+        // NOTE: we should fix https://github.com/dart-lang/gcloud/issues/23
+        //       and remove the toDatastoreKey conversion here.
+        final key =
+            _db.modelDB.toDatastoreKey(_db.emptyKey.append(User, id: lastId));
+        if (lastId != null) {
+          query.filter('__key__ >', key);
+        }
+      }
+
+      final users = await query.run().toList();
+      // We may return a page with users less then a limit, but we always
+      // set the continuation token to the correct value.
+      final newContinuationToken = users.length < limit
+          ? null
+          : _continuationCodec.encode(users.last.userId);
+      users.removeWhere((u) => u.isDeleted);
+
+      return api.AdminListUsersResponse(
+        users: users
+            .map(
+              (u) => api.AdminUserEntry(
+                userId: u.userId,
+                email: u.email,
+                oauthUserId: u.oauthUserId,
+              ),
+            )
+            .toList(),
+        continuationToken: newContinuationToken,
+      );
+    });
+  }
+
+  /// Removes user from the Datastore and updates the packages and other
+  /// entities they may have controlled.
+  Future removeUser(String userId) async {
+    await _withAdmin((u) async {
+      // TODO: remove once [withAuthenticatedUser] is using [User]
+      final user = await accountBackend.lookupUserById(userId);
+      if (user == null) return;
+      if (user.isDeleted) return;
+
+      _logger.info(
+          '${u.userId} (${u.email}) initiated the delete of ${user.userId} (${user.email})');
+
+      // Package.uploaders
+      final pkgQuery = _db.query<Package>()..filter('uploaders =', user.userId);
+      await for (final p in pkgQuery.run()) {
+        await _removeUploaderFromPackage(p.key, user.userId);
+      }
+
+      // PublisherMember
+      // Publisher.contactEmail
+      final memberQuery = _db.query<PublisherMember>()
+        ..filter('userId =', user.userId);
+      await for (final m in memberQuery.run()) {
+        await _removeMember(user, m);
+      }
+
+      // OAuthUserID
+      if (user.oauthUserId != null) {
+        final key = _db.emptyKey.append(OAuthUserID, id: user.oauthUserId);
+        final mapping = (await _db.lookup<OAuthUserID>([key])).single;
+        if (mapping != null) {
+          await _db.commit(deletes: [mapping.key]);
+        }
+      }
+
+      // User
+      await _markUserDeleted(user);
```

Comment:
Let's do this in the same transaction where we removed the `OAuthUserID` entity.